clang/lib/Driver/ToolChains/Flang.cpp

```diff
@@ -142,6 +142,26 @@ void Flang::addCodegenOptions(const ArgList &Args,
   if (shouldLoopVersion(Args))
     CmdArgs.push_back("-fversion-loops-for-stride");
 
+  Arg *aliasAnalysis = Args.getLastArg(options::OPT_falias_analysis,
+                                       options::OPT_fno_alias_analysis);
+  Arg *optLevel =
+      Args.getLastArg(options::OPT_Ofast, options::OPT_O, options::OPT_O4);
```

Comment:
What about other opt levels? Do we enable or disable alias analysis for these opt levels?

Reply:
It would be disabled unless `-falias-analysis` is specified directly.

What I am trying to accomplish here is for the frontend driver to usually just do what you expect, without having to remember to use this option. But I want there to still be a separate flag available to override this default behavior.

Reply:
I suppose alias analysis could make sense at `-Os` too, because it could enable better common sub-expression elimination and hoisting. What do you think?

Reply:
It looks like `clang` generates `tbaa` metadata at all opt levels, except `-O0`.  I think this makes sense: the optimization themselves need to decide how to use it, e.g. for improving performance/code-size/etc.

Reply:
> But I want there to still be a separate flag available to override this default behavior.

That's fine, but then one has to decide whether `-f{no}-alias-analysis` overrides `-O{n}` or not? I think that "explicit" request from a user should always take precedence. This leads to (pseudo code):
```
opts.AliasAnalysis = 0;
if (opt level requiring alias analysis)
  opts.AliasAnalysis  = 1;

// User request takes precedence when it comes to alias analysis.
if (-falias-analysis or -fno-alias-analysis) then
  "do whatever the user requested"
```

Separately, could you check what Clang does and make sure that that would be consistent?

Reply:
It turns out flang doesn't support `-Os`. Probably a bug here https://github.com/llvm/llvm-project/blob/main/flang/lib/Frontend/CompilerInvocation.cpp#L110

Reply:
Some discussion regarding `-Os` is here.
https://discourse.llvm.org/t/code-size-optimization-flags-in-flang/69482

Reply:
> That's fine, but then one has to decide whether -f{no}-alias-analysis overrides -O{n} or not? I think that "explicit" request from a user should always take precedence. This leads to (pseudo code):
>
> opts.AliasAnalysis = 0;
> if (opt level requiring alias analysis)
>  opts.AliasAnalysis  = 1;
> 
> / / User request takes precedence when it comes to alias analysis.
> if (-falias-analysis or -fno-alias-analysis) then
>  "do whatever the user requested"
>Separately, could you check what Clang does and make sure that that would be consistent?

@banach-space This is exactly the handling in the front-end driver as given below (and in `lib/Frontend/CompilerInvocation`). The flang driver is only deciding whether to forward or not.

```
  opts.AliasAnalysis = opts.OptimizationLevel > 0;
  if (auto *arg =
          args.getLastArg(clang::driver::options::OPT_falias_analysis,
                          clang::driver::options::OPT_fno_alias_analysis))
    opts.AliasAnalysis =
        arg->getOption().matches(clang::driver::options::OPT_falias_analysis);
```